packages/components/src/public/files/TemplateDownloadButton.tsx

```diff
@@ -1,155 +1,184 @@
+    const text = props.text ?? (isGridRenderer ? 'Download' : 'Template');
+    const [customTemplates, setCustomTemplates] = useState<ImportTemplate[]>([]);
+    const [downloading, setDownloading] = useState<boolean>(false);
+    const [loadingTemplates, setLoadingTemplates] = useState<LoadingState>(LoadingState.INITIALIZED);
     const { container, moduleContext } = useServerContext();
-    const homeFolderPath = getAppHomeFolderPath(container, moduleContext);
-
     const { api } = useAppContext();
+    const isLoaded = loadingTemplates === LoadingState.LOADED || !schemaQuery;
+    const isLoading = loadingTemplates === LoadingState.LOADING;
+    const schemaName = schemaQuery?.schemaName;
+    const queryName = schemaQuery?.queryName;
+    const hasTemplates = customTemplates.length > 0;
+    const isDownloadingOrLoading = downloading || isLoading;
 
-    useEffect(() => {
-        if (!schemaQuery || isGridRenderer || customTemplates) return;
-
-        (async () => {
-            await loadTemplates();
-        })();
-    }, [schemaQuery, isGridRenderer, customTemplates]);
+    const loadCustomTemplates = useCallback(async (): Promise<boolean> => {
+        let hasTemplates_ = false;
 
-    const loadTemplates = useCallback(async (): Promise<boolean> => {
         try {
-            setLoadingTemplates(true);
+            setLoadingTemplates(LoadingState.LOADING);
             const queryInfo = await api.query.getQueryDetails({
-                schemaName: schemaQuery.schemaName,
-                queryName: schemaQuery.queryName,
-                containerPath: homeFolderPath
+                containerPath: getAppHomeFolderPath(container, moduleContext),
+                schemaName,
+                queryName,
             });
-            const customTemplates_ = queryInfo.getCustomTemplates();
-            setCustomTemplates(customTemplates_);
-            return customTemplates_?.length > 0;
-        } catch (reason) {
-            console.error(reason);
+
+            const templates = queryInfo.getCustomTemplates() ?? [];
+            hasTemplates_ = templates.length > 0;
+
+            setCustomTemplates(templates);
+        } catch (error) {
+            console.error('Failed to load custom templates', error);
         } finally {
-            setLoadingTemplates(false);
+            setLoadingTemplates(LoadingState.LOADED);
         }
-    }, [schemaQuery, setLoadingTemplates, setCustomTemplates, homeFolderPath]);
 
-    const showDropdown = useMemo(() => {
-        return customTemplates?.length > 0 || (isGridRenderer && !customTemplates);
-    }, [customTemplates, isGridRenderer]);
+        return hasTemplates_;
+    }, [api.query, container, moduleContext, schemaName, queryName]);
+
+    const downloadCustomTemplate = useCallback(
+        (customTemplateUrl: string) => {
+            return async () => {
+                if (downloading) return;
+                setDownloading(true);
```

Comment:
I find the use of checking downloading and setting setDownloading twice weird. setDownloading hook is async so we can't guarantee` setDownloading(true);` is executed before `await downloadAttachment(customTemplateUrl);` and before `setDownloading(false);`. Am I not understanding this correctly? Should we not prevState.downloading instead?

Reply:
I've slightly updated the logic here -- also removed the async download of custom templates and just went back to having them be linked too.